File: backend/services/userop_srv/handler/message.py
import grpc
from loguru import logger

from userop_srv.model.models import LeaveMessage
from userop_srv.proto import message_pb2, message_pb2_grpc


class MessageServicer(message_pb2_grpc.MessageServicer):

    @logger.catch
    def MessageList(
        self,
        request: message_pb2.MessageRequest,
        context: grpc.ServicerContext
    ) -> message_pb2.MessageListResponse:
        """
        获取留言列表
        :param request: MessageRequest
        :param context: grpc.ServicerContext
        :return: MessageListResponse
        """
        rsp = message_pb2.MessageListResponse()
        messages = LeaveMessage.select()
        if request.userId:
            messages = messages.where(LeaveMessage.user == request.userId)

        rsp.total = messages.count()
        for msg in messages:
            msg_rsp = message_pb2.MessageResponse()

            msg_rsp.id = msg.id
            msg_rsp.userId = msg.user
            msg_rsp.messageType = msg.message_type
            msg_rsp.subject = msg.subject
            msg_rsp.message = msg.message
            msg_rsp.file = msg.file

            rsp.data.append(msg_rsp)

        return rsp

    @logger.catch
    def CreateMessage(
        self,
        request: message_pb2.MessageRequest,
        context: grpc.ServicerContext
    ) -> message_pb2.MessageResponse:
        """
        创建留言
        :param request: MessageRequest
        :param context: grpc.ServicerContext
        :return: MessageResponse
        """
        message = LeaveMessage(
            user=request.userId,
            message_type=request.messageType,
            subject=request.subject,
            message=request.message,
            file=request.file
        )
        message.save()

        return message_pb2.MessageResponse(
            id=message.id,
            messageType=message.message_type,
            subject=message.subject,
            message=message.message,
            file=message.file
        )

File: backend/apis/userop_web/api/message/message.go
package message

import (
	"apis/userop_web/api"
	"apis/userop_web/forms"
	"apis/userop_web/global"
	"apis/userop_web/models"
	"apis/userop_web/proto"
	"context"
	"net/http"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

func List(c *gin.Context) {
	request := &proto.MessageRequest{}

	userId, _ := c.Get("userId")
	claims := c.MustGet("claims").(*models.CustomClaims)
	if claims.AuthorityId == 1 {
		request.UserId = int32(userId.(uint))
	}
	rsp, err := global.MessageClient.MessageList(context.Background(), request)
	if err != nil {
		zap.S().Errorw("[List] 查询消息列表失败")
		api.HandleGrpcErrorToHttp(err, c)
		return
	}

	reMap := gin.H{
		"total": rsp.Total,
	}
	result := make([]interface{}, 0)
	for _, value := range rsp.Data {
		result = append(result, gin.H{
			"id":          value.Id,
			"user_id":     value.UserId,
			"messageType": value.MessageType,
			"subject":     value.Subject,
			"message":     value.Message,
			"file":        value.File,
		})
	}
	reMap["data"] = result

	c.JSON(http.StatusOK, reMap)
}

func New(c *gin.Context) {
	userId, _ := c.Get("userId")

	messageForm := forms.MessageForm{}
	if err := c.ShouldBindJSON(&messageForm); err != nil {
		api.HandleValidatorError(c, err)
		return
	}

	rsp, err := global.MessageClient.CreateMessage(context.Background(), &proto.MessageRequest{
		UserId:      int32(userId.(uint)),
		MessageType: messageForm.MessageType,
		Subject:     messageForm.Subject,
		Message:     messageForm.Message,
		File:        messageForm.File,
	})
	if err != nil {
		zap.S().Errorw("[New] 新建消息失败")
		api.HandleGrpcErrorToHttp(err, c)
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"id": rsp.Id,
	})
}
